app/javascript/src/Common/ajaxEvents.ts

```diff
@@ -1,26 +1,66 @@
+ * Set up AJAX event global listeners for Rails UJS (remote forms and links). They need to be
+ * declared with jQuery so that they can be overridden by local listeners (e.stopPropagation).
+ * Otherwise they will run after any other listeners, possibly breaking the UX.
+ * Example:
+ *   document.body.addEventListener('ajax:error', (event) => {
+ *     // Do stuff to local component
+ *
+ *     // Prevent global listener to be called
+ *     event.stopPropagation()
+ *   })
+ *
+ * Origin: app/assets/javascripts/ajax_events.js
  */
 
-const jQuery1 = window.$
-const expected = '1.11.3'
-const actual = jQuery1().jquery
+import $ from 'jquery'
 
-if (actual !== expected) {
-  throw new Error(`AJAX events declared in the wrong $. Expected: ${expected}, actual: ${actual}`)
-}
+import type { AJAXErrorEvent } from 'Types/rails-ujs'
 
-const ajaxSpinnerId = 'ajax-in-progress'
-
-jQuery1(document)
-  .on('ajax:before', () => {
-    $('body').append(`<div id="${ajaxSpinnerId}"><img src="/assets/ajax-loader.gif"></div>`)
+/**
+ * rails-ujs ajax events.
+ */
+$(document)
+  .on('ajax:before', (...args) => {
+    console.log('ajax:before on', args)
+    window.ThreeScale.showSpinner()
   })
-  .on('ajaxComplete ajax:complete', () => {
-    $(`#${ajaxSpinnerId}`).remove()
+  .on('ajax:beforeSend', (...args) => { console.log('ajax:beforeSend', args) })
+  .on('ajax:send', (...args) => { console.log('ajax:send on', args) })
+  .on('ajax:stopped', (...args) => { console.log('ajax:stopped', args) })
+  .on('ajax:success', (...args) => { console.log('ajax:success', args) })
+  .on('ajax:error', (event) => {
+    const [response, status] = (event.originalEvent as AJAXErrorEvent<unknown>).detail
+    console.error('ajax:error', response, status)
   })
-  .on('ajax:error', (_event, _xhr, status: string, error) => {
-    alert(`Request failed - ${status}`)
-    console.error(error)
+  .on('ajax:complete', (...args) => {
+    console.log('ajax:complete', args)
+    window.ThreeScale.hideSpinner()
+  })
+
+/**
+ * window.$.ajax events, triggered by Colorbox.
+ */
+window.$(document)
+  .on('ajaxStart', (...args) => { console.log('ajaxStart', args) })
+  .on('ajaxSend', (...args) => { console.log('ajaxSend', args) })
+  .on('ajaxSuccess', (...args) => { console.log('ajaxSuccess', args) })
+  .on('ajaxComplete', (...args) => { console.log('ajaxComplete', args) })
+  .on('ajaxError', (...args) => { console.log('ajaxError', args) })
+
+/**
+ * Handle errors in formtastic forms rendered inside a colorbox. The error template is passed a
+ * responseText inside the XHR response.
+ *
+ * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
+ * TODO: add a colorbox specific selector to separate this with non-legacy implementations.
+ */
+$(document).on('ajax:error', 'form[data-remote]:not(.pf-c-form)', (event) => {
+  const [,, xhr] = (event.originalEvent as AJAXErrorEvent<HTMLDocument>).detail
```

Comment:
```suggestion
  const [,, xhr] = (event.originalEvent as AJAXErrorEvent<Document>).detail
```

Just a TS suggestion...

```
/** @deprecated use Document */
interface HTMLDocument extends Document {
```